### ISSUE TITLE
Async virtual field in nested block sometimes does not work

### ISSUE THREAD

**Issue by yuriyyakym** [payloadcms/payload#5505]
### Describe the Bug

I have a nested structure with blocks field. One of my blocks has virtual reference field which I populate in `afterRead`.

The resulting promise of virtual field's `afterRead` hook is sometimes ignored.

The reason for that is the way how async traversal is implemented - it populates `fieldPromises` and `populationPromises` and then awaits them with `Promise.all`. But because population is itself async, some of nested structures push their promises too late (after `await Promise.all(fieldPromises)`. So even though `afterRead` hook is called, its result is ignored.

Bug is located [here](https://github.com/payloadcms/payload/blob/v2.11.2/packages/payload/src/fields/hooks/afterRead/index.ts#L73-L74)
```
  await Promise.all(fieldPromises)
  await Promise.all(populationPromises)
```

The easiest (but not perfect) solution would be something like:

```tsx
while (fieldPromises.length > 0 || populationPromises.length > 0) {
  await fieldPromises.shift()
  await populationPromises.shift()
}
```

or

```tsx
while (fieldPromises.length > 0 || populationPromises.length > 0) {
  const allPromises = [...fieldPromises, ...populationPromises]
  fieldPromises.splice(0)
  populationPromises.splice(0)
  await Promise.all(allPromises)
}
```

Since some of the promises are added too late, we can await them in the loop. We need to make sure that all the promises which were added to those arrays in current cycle are resolved.

Mutations are evil. Instead of pushing promises from nested fields to the root level array, parent promise could resolve only after its children's promises are resolved.

### To Reproduce

Create nested structure with blocks field that contains any async virtual field.
Note that `afterRead` hook has to have some delay `await wait(100)`, otherwise it will work in ~50% times.

Example structure:

```tsx
{
  name: 'blocks',
  type: 'blocks',
  defaultValue: [{ blockType: 'blockWithVirtualRelationField' }],
  blocks: [
    {
      slug: 'blockWithVirtualRelationField',
      fields: [
        {
          name: 'shouldPopulateVirtual',
          type: 'relationship',
          relationTo: relationsSlug,
          hasMany: true,
          hooks: {
            beforeChange: [
              ({ siblingData }) => {
                delete siblingData.shouldPopulateVirtual
              },
            ],
            afterRead: [
              async ({ req }) => {
                await wait(200)
                const { docs } = await req.payload.find({
                  collection: relationsSlug,
                })
                return docs
              },
            ],
          },
        },
      ],
    },
  ],
},
```

### Payload Version

2.11.2

### Proof

`all_categories` are sometimes an empty array.

https://github.com/payloadcms/payload/assets/18438911/d5629a16-19a7-4e7d-bd16-9f1edd63cb6e

**Comment by yuriyyakym** [comment]
I have created a PR which fixes the issue #5506. The fix is not ideal, but it works and shows where the problem lies.

**Comment by yuriyyakym** [comment]
For those experiencing experiencing the same issue, my temporary quick solution is to save a child's promise in context and then await it in parent's afterRead hook.

```tsx
// Child - Block virtual async field-level hook
afterRead: [
    async function ({ context, req: { payload } }) {
        const promise = payload
            .find({
                collection: 'example-category',
                limit: 10000,
                context: { ignoreFields: ['layout'] },
            })
            .then((a) => a.docs);

        context.promises = [promise];
        return promise;
    },
]
```

```tsx
// Parent - collection level hook

afterRead: [
    async ({ context }) => {
        await Promise.all((context as any).promises ?? []);
    },
],
```
